File: parser/src/sandbox-env.ts
export type SandboxEngineType = 'wasm' | 'firecracker' | 'chroot';

export interface SandboxInstance {
  id: string;
  engineType: SandboxEngineType;
  memoryMB: number;
  readOnlyFS: boolean;
  allowedCommands: string[];
  status: 'READY' | 'RUNNING' | 'TERMINATED';
  createdAt: string;
}

export interface SandboxExecutionResult {
  sandboxId: string;
  command: string;
  exitCode: number;
  stdout: string;
  stderr: string;
  executionTimeMs: number;
  isolated: boolean;
}

export class SandboxEnvEngine {
  private sandboxes: Map<string, SandboxInstance> = new Map();

  public createSandbox(
    id: string,
    engineType: SandboxEngineType = 'wasm',
    memoryMB: number = 128,
    readOnlyFS: boolean = true,
    allowedCommands: string[] = ['npm test', 'node index.js', 'python main.py']
  ): SandboxInstance {
    const instance: SandboxInstance = {
      id,
      engineType,
      memoryMB,
      readOnlyFS,
      allowedCommands,
      status: 'READY',
      createdAt: new Date().toISOString(),
    };

    this.sandboxes.set(id, instance);
    return instance;
  }

  public executeInSandbox(sandboxId: string, command: string, env?: Record<string, string>): SandboxExecutionResult {
    const instance = this.sandboxes.get(sandboxId);
    const startTime = Date.now();

    if (!instance) {
      return {
        sandboxId,
        command,
        exitCode: 1,
        stdout: '',
        stderr: `Error: Sandbox environment '${sandboxId}' not found.`,
        executionTimeMs: 0,
        isolated: false,
      };
    }

    const isAllowed = instance.allowedCommands.some((cmd) => command.startsWith(cmd.split(' ')[0]));
    if (!isAllowed) {
      return {
        sandboxId,
        command,
        exitCode: 126,
        stdout: '',
        stderr: `Permission Denied: Command '${command}' not in sandbox allowed_commands whitelist.`,
        executionTimeMs: Date.now() - startTime,
        isolated: true,
      };
    }

    instance.status = 'RUNNING';
    const executionTimeMs = Date.now() - startTime + 5; // Simulating micro-sandbox spinup & execution
    instance.status = 'READY';

    return {
      sandboxId,
      command,
      exitCode: 0,
      stdout: `[Wasm Sandbox Exec: ${instance.engineType}] ${command} executed cleanly (${instance.memoryMB}MB isolated stack)`,
      stderr: '',
      executionTimeMs,
      isolated: true,
    };
  }

  public verifySandboxIsolation(result: SandboxExecutionResult): boolean {
    return result.isolated && (result.exitCode === 0 || result.stderr.includes('Permission Denied'));
  }

  public getSandbox(id: string): SandboxInstance | undefined {
    return this.sandboxes.get(id);
  }
}

File: sdk/python/alp_sdk/sandbox_env.py
import time
from datetime import datetime, timezone
from typing import List, Dict, Any, Optional

class SandboxInstance:
    def __init__(
        self,
        sandbox_id: str,
        engine_type: str = "wasm",
        memory_mb: int = 128,
        read_only_fs: bool = True,
        allowed_commands: Optional[List[str]] = None,
        created_at: Optional[str] = None,
    ):
        self.id = sandbox_id
        self.engine_type = engine_type
        self.memory_mb = memory_mb
        self.read_only_fs = read_only_fs
        self.allowed_commands = allowed_commands or ["npm test", "node index.js", "python main.py"]
        self.status = "READY"
        self.created_at = created_at or datetime.now(timezone.utc).isoformat()

class SandboxExecutionResult:
    def __init__(
        self,
        sandbox_id: str,
        command: str,
        exit_code: int,
        stdout: str,
        stderr: str,
        execution_time_ms: float,
        isolated: bool = True,
    ):
        self.sandbox_id = sandbox_id
        self.command = command
        self.exit_code = exit_code
        self.stdout = stdout
        self.stderr = stderr
        self.execution_time_ms = execution_time_ms
        self.isolated = isolated

class SandboxEnvEngine:
    def __init__(self):
        self.sandboxes: Dict[str, SandboxInstance] = {}

    def create_sandbox(
        self,
        sandbox_id: str,
        engine_type: str = "wasm",
        memory_mb: int = 128,
        read_only_fs: bool = True,
        allowed_commands: Optional[List[str]] = None,
    ) -> SandboxInstance:
        instance = SandboxInstance(
            sandbox_id=sandbox_id,
            engine_type=engine_type,
            memory_mb=memory_mb,
            read_only_fs=read_only_fs,
            allowed_commands=allowed_commands,
        )
        self.sandboxes[sandbox_id] = instance
        return instance

    def execute_in_sandbox(self, sandbox_id: str, command: str) -> SandboxExecutionResult:
        start = time.time()
        instance = self.sandboxes.get(sandbox_id)

        if not instance:
            return SandboxExecutionResult(
                sandbox_id=sandbox_id,
                command=command,
                exit_code=1,
                stdout="",
                stderr=f"Error: Sandbox environment '{sandbox_id}' not found.",
                execution_time_ms=0.0,
                isolated=False,
            )

        cmd_root = command.split()[0]
        is_allowed = any(cmd.startswith(cmd_root) for cmd in instance.allowed_commands)

        if not is_allowed:
            return SandboxExecutionResult(
                sandbox_id=sandbox_id,
                command=command,
                exit_code=126,
                stdout="",
                stderr=f"Permission Denied: Command '{command}' not in sandbox allowed_commands whitelist.",
                execution_time_ms=(time.time() - start) * 1000,
                isolated=True,
            )

        return SandboxExecutionResult(
            sandbox_id=sandbox_id,
            command=command,
            exit_code=0,
            stdout=f"[Wasm Sandbox Exec: {instance.engine_type}] {command} executed cleanly",
            stderr="",
            execution_time_ms=(time.time() - start) * 1000 + 5.0,
            isolated=True,
        )

    def verify_sandbox_isolation(self, result: SandboxExecutionResult) -> bool:
        return result.isolated and (result.exit_code == 0 or "Permission Denied" in result.stderr)
